src/NuGet.Core/NuGet.Packaging/Rules/DependeciesGroupsForEachTFMRule.cs

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Xml.Linq;
+using NuGet.Client;
+using NuGet.Common;
+using NuGet.ContentModel;
+using NuGet.Frameworks;
+using NuGet.Packaging.Core;
+using NuGet.RuntimeModel;
+
+namespace NuGet.Packaging.Rules
+{
+    internal class DependeciesGroupsForEachTFMRule : IPackageRule
+    {
+        private static readonly NuGetFramework Net00 = NuGetFramework.Parse("net00");
+
+        public string MessageFormat => AnalysisResources.DependenciesGroupsForEachTFMHasNoExactMatch;
+
+        public string CompatMatchFoundWarningMessageFormat => AnalysisResources.DependenciesGroupsForEachTFMHasCompatMatch;
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
+        {
+            var files = builder.GetFiles().
+                Where(t => PathUtility.GetPathWithDirectorySeparator(t).Count(m => m == Path.DirectorySeparatorChar) > 1);
+            var packageNuspec = ExtractTFMsFromNuspec(builder.GetNuspec());
+            var (compatNotExactMatches, noExactMatchesFromFile, noExactMatchesFromNuspec) = CatagoriseTFMs(files, packageNuspec);
+            return GenerateWarnings(compatNotExactMatches, noExactMatchesFromFile, noExactMatchesFromNuspec);
+        }
+
+        internal IEnumerable<PackagingLogMessage> GenerateWarnings(HashSet<NuGetFramework> compatNotExactMatches, HashSet<NuGetFramework> noExactMatchesFromFile, HashSet<NuGetFramework> noExactMatchesFromNuspec)
+        {
+            (string noExactMatchString, string compatMatchString) = GenerateWarningString(noExactMatchesFromFile, noExactMatchesFromNuspec, (ICollection<NuGetFramework>)compatNotExactMatches);
+
+            var issues = new List<PackagingLogMessage>();
+
+            if (noExactMatchesFromFile.Count != 0 || noExactMatchesFromNuspec.Count != 0)
+            {
+                issues.Add(PackagingLogMessage.CreateWarning(string.Join("\n", MessageFormat, noExactMatchString),
+                    NuGetLogCode.NU5128));
+            }
+
+            if (compatNotExactMatches.Count != 0)
+            {
+                issues.Add(PackagingLogMessage.CreateWarning(string.Join("\n", CompatMatchFoundWarningMessageFormat, compatMatchString),
+                    NuGetLogCode.NU5130));
+            }
+
+            if (issues.Count != 0)
+            {
+                return issues;
+            }
+
+            return Array.Empty<PackagingLogMessage>();
```

Comment:
returning `Array.Empty<T>` is useful to avoid a `return new List<T>()`. But when an empty list already created, you can just return it. It avoids the need of the `if (issues.Count != 0)` condition so the method will be about 4 lines shorter. Only in the hyper-performance-sensitive scenarios you might want to ensure an empty list/array is reused rather, but in that case you'd probably also change the algorithm to not create the array unless it's actually needed.

Anyway, just FYI. Food for thought.

Reply:
Yeah just return issues regardless of how many elements the collection has.